File: client-sdk/ts-web/rt/src/callformat.ts
import * as oasis from '@oasisprotocol/client';
import * as deoxysii from '@oasisprotocol/deoxysii';

import * as mraeDeoxysii from './mrae/deoxysii';
import * as transaction from './transaction';
import * as types from './types';

/**
 * Call data key pair ID domain separation context base.
 */
export const CALL_DATA_KEY_PAIR_ID_CONTEXT_BASE = 'oasis-runtime-sdk/private: tx';

/**
 *  EncodeConfig is call encoding configuration.
 *  golang: oasis-sdk/client-sdk/go/callformat/callformat.go
 *  rust:
 */
export interface EncodeConfig {
    /**
     * publicKey is an optional runtime's call data public key to use for encrypted call formats.
     */
    publicKey?: types.KeyManagerSignedPublicKey;
    /**
     * epoch is the epoch of the ephemeral runtime key (when publicKey is set).
     */
    epoch?: oasis.types.longnum;
}

export interface MetaEncryptedX25519DeoxysII {
    sk: CryptoKey;
    pk: Uint8Array;
}

/**
 * encodeCallWithNonceAndKeys encodes a call based on its configured call format.
 * It returns the encoded call and any metadata needed to successfully decode the result.
 */
export async function encodeCallWithNonceAndKeys(
    nonce: Uint8Array,
    clientKP: CryptoKeyPair,
    call: types.Call,
    format: types.CallFormat,
    config?: EncodeConfig,
): Promise<[types.Call, unknown]> {
    switch (format) {
        case transaction.CALLFORMAT_PLAIN:
            return [call, undefined];
        case transaction.CALLFORMAT_ENCRYPTED_X25519DEOXYSII:
            if (config?.publicKey === undefined) {
                throw new Error('callformat: runtime call data public key not set');
            }
            const pk = await mraeDeoxysii.publicKeyFromKeyPair(clientKP);
            const sk = clientKP.privateKey;
            const rawCall = oasis.misc.toCBOR(call);
            const zeroBuffer = new Uint8Array(0);
            const sealedCall = await mraeDeoxysii.boxSeal(
                nonce,
                rawCall,
                zeroBuffer,
                config.publicKey.key,
                sk,
            );
            const envelope: types.CallEnvelopeX25519DeoxysII = {
                pk: pk,
                nonce: nonce,
                data: sealedCall,
            };
            if (config.epoch) {
                envelope.epoch = config.epoch;
            }
            const encoded: types.Call = {
                format: transaction.CALLFORMAT_ENCRYPTED_X25519DEOXYSII,
                body: envelope,
            };
            const meta: MetaEncryptedX25519DeoxysII = {
                sk: sk,
                pk: config.publicKey.key,
            };
            return [encoded, meta];
        default:
            throw new Error(`callformat: unsupported call format: ${format}`);
    }
}

/**
 * encodeCall randomly generates nonce and keyPair and then call encodeCallWithNonceAndKeys
 * It returns the encoded call and any metadata needed to successfully decode the result.
 */
export async function encodeCall(
    call: types.Call,
    format: types.CallFormat,
    config?: EncodeConfig,
): Promise<[types.Call, unknown]> {
    const nonce = new Uint8Array(deoxysii.NonceSize);
    crypto.getRandomValues(nonce);
    const clientKP = await mraeDeoxysii.generateKeyPair(true);
    return await encodeCallWithNonceAndKeys(nonce, clientKP, call, format, config);
}

/**
 * decodeResult performs result decoding based on the specified call format metadata.
 */
export async function decodeResult(
    result: types.CallResult,
    format: types.CallFormat,
    meta?: unknown,
): Promise<types.CallResult> {
    switch (format) {
        case transaction.CALLFORMAT_PLAIN:
            // In case of plain-text data format, we simply pass on the result unchanged.
            return result;
        case transaction.CALLFORMAT_ENCRYPTED_X25519DEOXYSII:
            if (result.unknown) {
                if (meta) {
                    const metaEncryptedX25519DeoxysII = meta as MetaEncryptedX25519DeoxysII;
                    const envelop = oasis.misc.fromCBOR(
                        result.unknown,
                    ) as types.ResultEnvelopeX25519DeoxysII;
                    const zeroBuffer = new Uint8Array(0);
                    const pt = await mraeDeoxysii.boxOpen(
                        envelop?.nonce,
                        envelop?.data,
                        zeroBuffer,
                        metaEncryptedX25519DeoxysII.pk,
                        metaEncryptedX25519DeoxysII.sk,
                    );
                    return oasis.misc.fromCBOR(pt) as types.CallResult;
                } else {
                    throw new Error(
                        `callformat: MetaEncryptedX25519DeoxysII data is required for callformat: CALLFORMAT_ENCRYPTED_X25519DEOXYSII`,
                    );
                }
            } else if (result.fail) {
                throw new Error(
                    `callformat: failed call: module: ${result.fail.module} code: ${result.fail.code} message: ${result.fail.message}`,
                );
            }
            throw Object.assign(new Error(`callformat: unexpected result: ${result.ok}`), {
                source: result,
            });
        default:
            throw new Error(`callformat: unsupported call format: ${format}`);
    }
}

File: client-sdk/go/callformat/callformat.go
package callformat

import (
	"crypto/rand"
	"fmt"

	"github.com/oasisprotocol/curve25519-voi/primitives/x25519"
	"github.com/oasisprotocol/deoxysii"
	"github.com/oasisprotocol/oasis-core/go/common/cbor"
	mraeDeoxysii "github.com/oasisprotocol/oasis-core/go/common/crypto/mrae/deoxysii"

	"github.com/oasisprotocol/oasis-sdk/client-sdk/go/types"
)

// EncodeConfig is call encoding configuration.
type EncodeConfig struct {
	// PublicKey is an optional runtime's call data public key to use for encrypted call formats.
	PublicKey *types.SignedPublicKey
	// Epoch is the epoch of the ephemeral runtime key (when PublicKey is set).
	Epoch uint64
}

type metaEncryptedX25519DeoxysII struct {
	// sk is the ephemeral secret key for X25519.
	sk *x25519.PrivateKey
	// pk is the current calldata X25519 public key.
	pk *x25519.PublicKey
}

func encodeCallEncryptedX25519DeoxysII(call *types.Call, pk *x25519.PublicKey, sk *x25519.PrivateKey, nonce [deoxysii.NonceSize]byte, cfg *EncodeConfig) (*types.Call, *metaEncryptedX25519DeoxysII) {
	// Seal serialized plain call.
	rawCall := cbor.Marshal(call)
	sealedCall := mraeDeoxysii.Box.Seal(nil, nonce[:], rawCall, nil, &cfg.PublicKey.PublicKey, sk)

	encoded := &types.Call{
		Format: types.CallFormatEncryptedX25519DeoxysII,
		Method: "",
		Body: cbor.Marshal(&types.CallEnvelopeX25519DeoxysII{
			Pk:    *pk,
			Nonce: nonce,
			Epoch: cfg.Epoch,
			Data:  sealedCall,
		}),
		ReadOnly: call.ReadOnly,
	}
	meta := &metaEncryptedX25519DeoxysII{
		sk: sk,
		pk: &cfg.PublicKey.PublicKey,
	}
	return encoded, meta
}

// EncodeCall encodes a call based on its configured call format.
//
// It returns the encoded call and any metadata needed to successfully decode the result.
func EncodeCall(call *types.Call, cf types.CallFormat, cfg *EncodeConfig) (*types.Call, interface{}, error) {
	switch cf {
	case types.CallFormatPlain:
		// In case of the plain-text data format, we simply pass on the call unchanged.
		return call, nil, nil
	case types.CallFormatEncryptedX25519DeoxysII:
		// We require the runtime's call data public key to be configured.
		if cfg == nil || cfg.PublicKey == nil {
			return nil, nil, fmt.Errorf("callformat: runtime call data public key not set")
		}

		// Generate ephemeral X25519 key pair.
		pk, sk, err := x25519.GenerateKey(rand.Reader)
		if err != nil {
			return nil, nil, fmt.Errorf("callformat: failed to generate ephemeral X25519 key pair: %w", err)
		}
		// Generate random nonce.
		var nonce [deoxysii.NonceSize]byte
		if _, err := rand.Read(nonce[:]); err != nil {
			return nil, nil, fmt.Errorf("callformat: failed to generate random nonce: %w", err)
		}

		encoded, meta := encodeCallEncryptedX25519DeoxysII(call, pk, sk, nonce, cfg)
		return encoded, meta, nil
	default:
		return nil, nil, fmt.Errorf("callformat: unsupported call format: %s", cf)
	}
}

// DecodeResult performs result decoding based on the specified call format metadata.
func DecodeResult(result *types.CallResult, meta interface{}) (*types.CallResult, error) {
	switch m := meta.(type) {
	case nil:
		// In case of plain-text data format, we simply pass on the result unchanged.
		return result, nil
	case *metaEncryptedX25519DeoxysII:
		var envelope types.ResultEnvelopeX25519DeoxysII
		switch {
		case result.IsUnknown():
			if err := cbor.Unmarshal(result.Unknown, &envelope); err != nil {
				return nil, fmt.Errorf("callformat: malformed result envelope: %w", err)
			}
		case result.IsSuccess():
			if err := cbor.Unmarshal(result.Ok, &envelope); err != nil {
				return nil, fmt.Errorf("callformat: malformed result envelope: %w", err)
			}
		default:
			// Submission could fail before call format processing so the result would be plain.
			return nil, result.Failed
		}

		// Open sealed envelope.
		var (
			pt  []byte
			err error
		)
		if pt, err = mraeDeoxysii.Box.Open(nil, envelope.Nonce[:], envelope.Data, nil, m.pk, m.sk); err != nil {
			return nil, fmt.Errorf("callformat: failed to open result envelope: %w", err)
		}

		var output types.CallResult
		if err = cbor.Unmarshal(pt, &output); err != nil {
			return nil, fmt.Errorf("callformat: malformed result: %w", err)
		}
		return &output, nil
	default:
		return nil, fmt.Errorf("callformat: unsupported call format: %T", m)
	}
}
